File: AI/assgn1 - Search Algos/bestfs.py
import copy
import math
from collections import deque

init_config = [
    ['8', '4', '7'],
    ['1', '6', '5'],
    ['2', '3', ' ']
]

goal = [
    ['1', '8', '7'],
    ['2', ' ', '6'],
    ['3', '4', '5']
]

def main():
    BestFS(init_config)

def BestFS(config):
    queue = deque()
    queue.append([config])
    
    visited = []
    visited.append(config)
    
    while (len(queue) != 0):
        path = queue.popleft()
        parent = copy.deepcopy(path[len(path) - 1])

        # prepare to extend
        neighbours = get_neighours(parent)
        
        scores = []
        for neighbour in neighbours:
            scores.append(calc_score(neighbour))
        
        # sort according to heuristic scores
        sorted_neighbours = [x for _,x in sorted(zip(scores, neighbours))]

        # check if any neighbour is the goal, if not then extend
        for neighbour in sorted_neighbours:
            if neighbour == goal:
                print("Goal found at depth: " + str(len(path)))
                print(path + goal)
                return len(path)
            elif neighbour not in visited:
                queue.append(path + [neighbour])
                visited.append(neighbour)
    
    # couldnt find goal
    print("Could not find goal.")
    return -1

# computes all possible mutations of given config
def get_neighours(config):
    mutations = []
    for i in range(3):
        if ' ' in config[i]:
            j = config[i].index(' ')
            new_moves = [
                (i - 1, j), (i + 1, j),
                (i, j + 1), (i, j - 1)
            ]

            for a, b in new_moves:
                if (a >= 0 and b >= 0) and (a < 3 and b < 3):
                    temp = copy.deepcopy(config)
                    temp[i][j], temp[a][b] = temp[a][b], temp[i][j]
                    mutations.append(temp)
            break
    
    return mutations

# heuristic function
def calc_score(config):
   #return tiles_in_place(config)
   #return manhattan_distance(config)
   return euclidean_distance(config)

def g (config):
    return 0

def manhattan_distance(config):
    sum = 0

    for i in range(3):
        for j in range(3):
            for k in range(3):
                for l in range(3):
                    if goal[i][j] != ' ' and goal[i][j] == config[k][l]:
                        sum += (abs(i - k) + abs(j - l))

    return sum

def euclidean_distance(config):
    sum = 0

    for i in range(3):
        for j in range(3):
            for k in range(3):
                for l in range(3):
                    if goal[i][j] != ' ' and goal[i][j] == config[k][l]:
                        sum += math.sqrt((i - k)**2 + abs(j - l)**2)

    return sum

def tiles_in_place(config):
    score = 0
    for i in range(3):
        for j in range(3):
            if config[i][j] == goal[i][j]:
                score += 1

    return score

if __name__ == "__main__":
    main()

File: AI/assgn1 - Search Algos/bestfs.cpp
#include <iostream>
#include <vector>
#include <deque>
#include <cmath>
#include <algorithm>

using namespace std;

void print(vector<vector<char> > x)
{
	for (int i = 0; i < 3; i++)
	{
		for (int j = 0; j < 3; j++)
		{
			cout << x[i][j] << " ";
		}
		cout << "\n";
	}
}

vector<vector<char> > swap(vector<vector<char> > a, int i1, int j1, int i2, int j2)
{
	char temp = a[i1][j1];
	a[i1][j1] = a[i2][j2];
	a[i2][j2] = temp;
	
	return a;
}

vector<vector<vector<char> > > get_neighbours(vector<vector<char> > config)
{
	int i, j;
	bool found = false;
	for (i = 0; i < 3; i++)
	{
		for (j = 0; j < 3; j++)
		{
			if (config[i][j] == ' ')
			{
				found = true;
				break;
			}
		}
		if (found)
			break;
	}

	// calculate neighbours
	vector<vector<vector<char> > > neighbours;
	
	if (i - 1 >= 0)
	{	
		vector<vector<char> > temp = config;
		neighbours.push_back(swap(temp, i, j, i - 1, j));
		
	}
	if (i + 1 <= 2)
	{
		vector<vector<char> > temp = config;
		neighbours.push_back(swap(temp, i, j, i + 1, j));
	}
	if (j - 1 >= 0)
	{
		vector<vector<char> > temp = config;
		neighbours.push_back(swap(temp, i, j, i, j - 1));
	}
	if (j + 1 <= 2)
	{
		vector<vector<char> > temp = config;
		neighbours.push_back(swap(temp, i, j, i, j + 1));
	}

	return neighbours;
}

bool find(vector<vector<vector<char> > > arr, vector<vector<char> > to_find)
{
	for (vector<vector<char> > i : arr)
	{
		if (i == to_find)
			return true;
	}

	return false;
}

int euclidean_distance(vector<vector<char> > nodes, vector<vector<char> > goal)
{
	int sum = 0;
	
	for (int i = 0; i < 3; i++)
	{
		for (int j = 0; j < 3; j++)
		{
			for (int k = 0; k < 3; k++)
			{
				for (int l = 0; l < 3; l++)
				{
					if (goal[i][j] != ' ' && goal[i][j] == nodes[k][l])
						sum += sqrt((i - k)*(i - k) + (j - l)*(j - l));
				}
			}
		}
	}
	
	return sum;
}
// something is wrong
//int out_of_row_col(vector<vector<char> > nodes, vector<vector<char> > goal)
//{
//	int sum = 0;
//	
//	for (int i = 0; i < 3; i++)
//	{
//		for (int j = 0; j < 3; j++)
//		{
//			if (nodes[i][j] != ' ' && nodes[i][j] != goal[i][j])
//				sum += 1;
//		}
//	}
//	
//	for (int j = 0; j < 3; j++)
//	{
//		for (int i = 0; i < 3; i++)
//		{
//			if (nodes[i][j] != ' ' && nodes[i][j] != goal[i][j])
//				sum += 1;
//		}
//	}
//	
//	return sum;
//}

vector<pair<vector<vector<char> >, int> > get_scores(vector<vector<vector<char> > > nodes, vector<vector<char> > goal)
{
	vector<pair<vector<vector<char> >, int> > scores;
	for (auto node : nodes)
	{
		scores.push_back(make_pair(node, euclidean_distance(node, goal)));
	}
	
	return scores;
}

struct ordering {
    bool operator ()(pair<vector<vector<char> >, int> const& a, 
                     pair<vector<vector<char> >, int> const& b) {
        return a.second < b.second;
    }
};

int main(void)
{
	vector<vector<char> > initial_config = { {'8', '4', '7'},
												{'1', '6','5'},
												{'2', '3', ' '}
											};
	vector<vector<char> > goal = { {'1', '8', '7'},
									{'2', ' ', '6'},
									{'3', '4', '5'}
								};

	deque<vector<vector<vector<char> > > > queue;
	vector<vector<vector<char> > > visited;
	queue.push_back({ initial_config });

	while (queue.size() != 0)
	{
		vector<vector<vector<char> > > path = queue.front();
		queue.pop_front();

		vector<vector<char> > last = path[path.size() - 1];

		// if the neighbour is the goal
		if (last == goal)
		{
			cout << "Path to goal:\n";
			for (vector<vector<char> > x : path)
			{
				print(x);
				cout << "\n";
			}
			return 0;
		}

		// get neighbours
		vector<vector<vector<char> > > neighbours = get_neighbours(last);
		
		// sort neighbours
		vector<pair<vector<vector<char> >, int> > scores = get_scores(neighbours, goal);
		sort(scores.begin(), scores.end(), ordering());

		vector<vector<vector<char> > > sorted_neighbours;
		for (int i = 0; i < scores.size(); i++)
		{
			sorted_neighbours.push_back(scores[i].first);
		}
	
		for (int c = 0; c < sorted_neighbours.size(); c++)
		{
			// extend that neighbour
			if (!find(visited, sorted_neighbours[c]))
			{
				vector<vector<vector<char> > > temp = path;
				temp.push_back(sorted_neighbours[c]);
				queue.push_back(temp);
				visited.push_back(sorted_neighbours[c]);
			}
		}
	}

	cout << "Could not find goal\n";

	return 0;
}
